rustworkx-core/src/token_swapper.rs

```diff
@@ -0,0 +1,608 @@
+// not use this file except in compliance with the License. You may obtain
+// a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+// WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+// License for the specific language governing permissions and limitations
+// under the License.
+
+use rand::distributions::{Standard, Uniform};
+use rand::prelude::*;
+use rand_pcg::Pcg64;
+use std::hash::Hash;
+
+use hashbrown::HashMap;
+use petgraph::stable_graph::{NodeIndex, StableGraph};
+use petgraph::visit::{
+    EdgeCount, GraphBase, IntoEdges, IntoNeighborsDirected, IntoNodeIdentifiers, NodeCount,
+    NodeIndexable, Visitable,
+};
+use petgraph::Directed;
+use petgraph::Direction::{Incoming, Outgoing};
+use rayon_cond::CondIterator;
+
+use crate::connectivity::find_cycle;
+use crate::dictmap::*;
+use crate::shortest_path::dijkstra;
+use crate::traversal::dfs_edges;
+
+type Swap = (NodeIndex, NodeIndex);
+type Edge = (NodeIndex, NodeIndex);
+
+struct TokenSwapper<G: GraphBase>
+where
+    G::NodeId: Eq + Hash,
+{
+    // The input graph
+    graph: G,
+    // The user-supplied mapping to use for swapping tokens
+    mapping: HashMap<G::NodeId, G::NodeId>,
+    // Number of trials
+    trials: usize,
+    // Seed for random selection of a node for a trial
+    seed: Option<u64>,
+    // Threshold for how many nodes will trigger parallel iterator
+    parallel_threshold: usize,
+    // Map of NodeId to NodeIndex
+    node_map: HashMap<G::NodeId, NodeIndex>,
+    // Map of NodeIndex to NodeId
+    rev_node_map: HashMap<NodeIndex, G::NodeId>,
+}
+
+impl<G> TokenSwapper<G>
+where
+    G: NodeCount
+        + EdgeCount
+        + IntoEdges
+        + Visitable
+        + NodeIndexable
+        + IntoNeighborsDirected
+        + IntoNodeIdentifiers
+        + Send
+        + Sync,
+    G::NodeId: Hash + Eq + Send + Sync,
+{
+    fn new(
+        graph: G,
+        mapping: HashMap<G::NodeId, G::NodeId>,
+        trials: Option<usize>,
+        seed: Option<u64>,
+        parallel_threshold: Option<usize>,
+    ) -> Self {
+        TokenSwapper {
+            graph,
+            mapping,
+            trials: trials.unwrap_or(4),
+            seed,
+            parallel_threshold: parallel_threshold.unwrap_or(50),
+            node_map: HashMap::with_capacity(graph.node_count()),
+            rev_node_map: HashMap::with_capacity(graph.node_count()),
+        }
+    }
+
+    fn map(&mut self) -> Vec<Swap> {
+        let num_nodes = self.graph.node_bound();
+        let num_edges = self.graph.edge_count();
+
+        // Directed graph with nodes matching ``graph`` and
+        // edges for neighbors closer than nodes
+        let mut digraph = StableGraph::with_capacity(num_nodes, num_edges);
```

Comment:
We should make this `self.graph.node_bound()` instead of `num_nodes()` because we're always going to overallocate the way this is being done since we grow the graph to `node_bound` no matter what.

Reply:
Actually if you look above this, I actually changed `num_nodes` to be `node_bound`. Should have changed the var name.

Reply:
Ah ok, yeah the variable name tripped me up. If this is already using the upper bound on the node index we're good here.